dev-tools/harness.js

```diff
@@ -0,0 +1,247 @@
+		if (this.input) console.log(message);
+		this.rawInputLog.push(message);
+		super._write(message);
+	}
+}
+
+class MultiRunner {
+	constructor(options) {
+		this.options = Object.assign({}, options);
+
+		this.totalGames = options.totalGames;
+
+		this.prng = (options.prng && !Array.isArray(options.prng)) ?
+			options.prng : new PRNG(options.prng);
+		this.options.prng = this.prng;
+
+		this.format = options.format;
+		this.cycle = !!options.cycle;
+		this.all = !!options.all;
+
+		this.async = !!options.async;
+
+		this.formatIndex = 0;
+		this.numGames = 0;
+	}
+
+	async run() {
+		let games = [];
+		let format, lastFormat;
+		let failures = 0;
+		while ((format = this.getNextFormat())) {
+			if (this.all && lastFormat && format !== lastFormat) {
+				if (this.async) await Promise.all(games);
+				games = [];
+			}
+
+			const seed = this.prng.seed;
+			const game = new Runner(Object.assign({format}, this.options)).run().catch(err => {
+				failures++;
+				console.error(
+					`Run \`node dev-tools/harness 1 --format=${format} --seed=${seed.join()}\` ` +
+					`to debug (optionally with \`--output\` and/or \`--input\` for more info):\n`, err);
+			});
+
+			if (!this.async) await game;
+			games.push(game);
+			lastFormat = format;
+		}
+
+		if (this.async) await Promise.all(games);
+		return failures;
+	}
+
+	getNextFormat() {
+		if (this.formatIndex > FORMATS.length) return false;
+
+		if (this.numGames++ < this.totalGames) {
+			if (this.format) {
+				return this.format;
+			} else if (this.all) {
+				return FORMATS[this.formatIndex];
+			} else if (this.cycle) {
+				const format = FORMATS[this.formatIndex];
+				this.formatIndex = (this.formatIndex + 1) % FORMATS.length;
+				return format;
+			} else {
+				return this.prng.sample(FORMATS);
+			}
+		} else if (this.all) {
+			this.numGames = 1;
+			this.formatIndex++;
+			return FORMATS[this.formatIndex];
+		}
+
+		return false;
+	}
+}
+
+// Tracks whether some promises threw errors that weren't caught so we can log
+// and exit with a non-zero status to fail any tests. This "shouldn't happen"
+// because we're "great at propagating promises (TM)", but better safe than sorry.
+const RejectionTracker = new class {
+	constructor() {
+		this.unhandled = [];
+	}
+
+	onUnhandledRejection(reason, promise) {
+		this.unhandled.push({reason, promise});
+	}
+
+	onRejectionHandled(promise) {
+		this.unhandled.splice(this.unhandled.findIndex(u => u.promise === promise), 1);
+	}
+
+	onExit(code) {
+		let i = 0;
+		for (const u of this.unhandled) {
+			const error = (u.reason instanceof Error) ? u.reason :
+				new Error(`Promise rejected with value: ${u.reason}`);
+			console.error(error.stack);
+			i++;
+		}
+		process.exit(code + i);
+	}
+
+	register() {
+		process.on('unhandledRejection', (r, p) => RejectionTracker.onUnhandledRejection(r, p));
+		process.on('rejectionHandled', p => RejectionTracker.onRejectionHandled(p));
+		process.on('exit', c => RejectionTracker.onExit(c));
+	}
+}();
```

Comment:
Given that the ``register()`` method refers to a unique object, it makes more sense for ``RejectionTracker`` to be a singleton:

```js
const RejectionTracker = {
  unhandled: [],
  ...
};

Reply:
Good point.